### PR TITLE
[KYUUBI #5505][FLINK] Support HELP command

### PR DESCRIPTION
**Pull request by YesOrNo828**
### _Why are the changes needed?_

resolve: #5505 

Show available commands when users type 'HELP;' in the beeline.

#### Solution:

- Using `ExtendedParser` parse statement and return Operation of Flink engine.
- Check whether the operation is HelpOperation or not.  
- dependency on `flink-table-planner-loader.jar`.

#### **Why not using Flink SQL Client Parser(SqlCommandParserImpl) to obtain the Command enum?**
Flink 1.16's approach:
```
val opt:Optional[Operation] = org.apache.flink.table.client.cli.SqlCommandParserImpl.parseCommand()
check opt.get() instance of HelpOperation or not
if yes return CliStrings.MESSAGE_HELP
```

Flink 1.17 & 1.18
```
val opt: Optional[Command] = org.apache.flink.table.client.cli.parser.SqlCommandParserImpl.parseStatement()
check opt.get() is Command.HELP or not
if yes return CliStrings.MESSAGE_HELP
```

The `Command` class is added from Flink1.17;
The `SqlCommandParserImpl` package is changed, and the method name is changed from Flink1.17;

This approach requires distinguishing between different Flink versions and maintaining both implementations.
It's more complicated, so abandoned.

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/master/contributing/code/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/ExecuteStatement.scala [hunk 1]
```diff
@@ -59,16 +66,49 @@ class ExecuteStatement(
...
+        return
+      }
+
       val resultFetcher = executor.executeStatement(
         new OperationHandle(getHandle.identifier),
         statement)
       jobId = FlinkEngineUtils.getResultJobId(resultFetcher)
-      resultSet = ResultSetUtil.fromResultFetcher(resultFetcher, resultMaxRows, resultFetchTimeout)
+      resultSet = ResultSetUtil
+        .fromResultFetcher(resultFetcher, resultMaxRows, resultFetchTimeout)
```
Comment (pan3793) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/ExecuteStatement.scala:82]: unnecessary change?

File: pom.xml [hunk 2]
```diff
@@ -1385,6 +1385,13 @@
                 <scope>provided</scope>
             </dependency>
 
+            <dependency>
+                <groupId>org.apache.flink</groupId>
+                <artifactId>flink-table-planner-loader</artifactId>
+                <version>${flink.version}</version>
+                <scope>provided</scope>
```
Comment (pan3793) [inline pom.xml:1392]: it's not intended to define scope in the root pom dependencyManagement block

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/ExecuteStatement.scala [hunk 3]
```diff
@@ -59,16 +66,49 @@ class ExecuteStatement(
...
+        .impl("org.apache.flink.table.planner.parse.ExtendedParser")
+        .build().newInstance()
+    invokeAs(extendedParser, "parse", (classOf[String], statement))
+      .asInstanceOf[Optional[Operation]]
+  }
+
+  private def getPlannerModuleClassLoader: ClassLoader = {
+    try {
+      val plannerModule =
+        invokeAs("org.apache.flink.table.planner.loader.PlannerModule", "getInstance")
```
Comment (pan3793) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/ExecuteStatement.scala:105]: could this be invoked directly?

Comment (pan3793) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/ExecuteStatement.scala:105]: it's encouraged to use `Dyn*` for reflection in Flink engine module

Comment (YesOrNo828) [reply externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/ExecuteStatement.scala:105]: `PlannerModule` is package-private, we can't invoke it directly here.

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/ExecuteStatement.scala [hunk 4]
```diff
@@ -71,4 +85,33 @@ class ExecuteStatement(
...
+        .hiddenImpl("org.apache.flink.table.planner.loader.PlannerModule")
+        .buildStaticChecked()
+        .invoke().asInstanceOf[AnyRef]
+
+      DynFields.builder()
+        .hiddenImpl(plannerModule.getClass, "submoduleClassLoader")
+        .build[ClassLoader].bind(plannerModule).get
+    } catch {
+      case e: Exception =>
+        throw new TableException("Error obtaining Flink planner module ClassLoader", e)
```
Comment (pan3793) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/ExecuteStatement.scala:114]: Could we add some messages to guide users on how to fix it?

**Review by pan3793 — APPROVED** [review]
LGTM, except to one minor comment

File: externals/kyuubi-flink-sql-engine/src/test/scala/org/apache/kyuubi/engine/flink/operation/FlinkOperationSuite.scala [hunk 5]
```diff
@@ -1265,4 +1266,14 @@ abstract class FlinkOperationSuite extends HiveJDBCTestHelper with WithFlinkTest
...
     assert(exception.getMessage === "Futures timed out after [60000 milliseconds]")
   }
+
+  test("execute statement - help") {
+    withJdbcStatement() { stmt =>
+      val resultSet = stmt.executeQuery("help")
+      val metadata = resultSet.getMetaData
+      assert(metadata.getColumnName(1) === "result")
+      assert(resultSet.next())
+      assert(resultSet.getString(1).equals(CliStrings.MESSAGE_HELP.toString))
```
Comment (link3280) [inline externals/kyuubi-flink-sql-engine/src/test/scala/org/apache/kyuubi/engine/flink/operation/FlinkOperationSuite.scala:1276]: IIUC, the command `CLEAR` and `QUIT/EXIT` in the help message are not supported. Would that be kind of misleading?

**Review by link3280 — COMMENTED** [review]
Thanks for the PR! I left some comments please kindly take a look. @YesOrNo828

File: externals/kyuubi-flink-sql-engine/src/test/scala/org/apache/kyuubi/engine/flink/operation/FlinkOperationSuite.scala (hunk 5 above)
Comment (YesOrNo828) [reply externals/kyuubi-flink-sql-engine/src/test/scala/org/apache/kyuubi/engine/flink/operation/FlinkOperationSuite.scala:1276]: Yes, I think you're right. This `CliStrings.MESSAGE_HELP` message is described in the Flink SQL Client. CLEAR and QUIT/EXIT are not supported for Beeline.

How about we define the help message in the Flink Engine? Removing the CLEAR and QUIT/EXIT command descriptions may be a little tricky from `CliStrng.MESSAGE_HELP`. We could also add some new descriptions, for example: show jobs, stop job 'job_id', create Table, drop table, etc. WDYT?

Comment (link3280) [reply externals/kyuubi-flink-sql-engine/src/test/scala/org/apache/kyuubi/engine/flink/operation/FlinkOperationSuite.scala:1276]: That sounds promising. +1 from my side.

Comment (YesOrNo828) [reply externals/kyuubi-flink-sql-engine/src/test/scala/org/apache/kyuubi/engine/flink/operation/FlinkOperationSuite.scala]: This is the screenshot of `HELP;` command. Do you have a time to take a look? @link3280 

image

Comment (link3280) [reply externals/kyuubi-flink-sql-engine/src/test/scala/org/apache/kyuubi/engine/flink/operation/FlinkOperationSuite.scala]: I think job statements are very common and should be included. Plus, we could state that the help message only contains the most frequently used statements. Thanks!

**Comment by link3280** [comment]
@YesOrNo828 Merged. Thanks!
